### DISCUSSION TITLE
Vault dump from Lovecraft Country?

### DISCUSSION THREAD

**Discussion by noahgibbs** [ChatTheatre/Community#7] [General]
Hey, @saraht45! Chris mentions that you have a copy of the last Lovecraft Country code, including a statedump. He was hoping you could get a dump of the XML objects out of it.

@Shentino wrote a great object-dumping tool ("vault tool") that's checked into ChatTheatre/SkotOS. I've installed it on the live Gables server and it works there. It seems likely to work for dumping XML objects for Lovecraft Country, too. Documentation on using it is here: https://chattheatre.github.io/SkotOS-Doc/Exploring_SkotOS.html#dumping-lpc-objects-on-the-gables

Unfortunately there's no documentation in place on *installing* it. Fortunately, it's pretty simple. ChatTheatre/SkotOS has two relevant directories: [skoot/usr/System/sys/tool](https://github.com/ChatTheatre/SkotOS/tree/master/skoot/usr/System/sys/tool) and [skoot/usr/System/lib/tool](https://github.com/ChatTheatre/SkotOS/tree/master/skoot/usr/System/lib/tool). If you copy those five files into place in the same two directories and compile the tool from the developer wiztool interface ("cd /usr/System/sys/tool", "compile vault.c"), it should start working.

(I've just created a PR to add these same install instructions to the "Exploring SkotOS" docs.)

Sanity check: you have a dev account and you can get in on the telnet port, yes?

I *think* that with these instructions you should be able to get an object dump as XML files out of Lovecraft Country or similar old games. If you get compile errors or it won't dump objects, let me know and I might be able to help (and/or Shentino might.)

You *will* see some errors as it dumps, as shown on the Exploring_SkotOS page above. But assuming it basically works, you'll find the dumped XML files under skoot/data/vault in a lot of subdirectories.

**Comment by saraht45** [comment]
Cool! I'll give it a shot tomorrow. (Tonight if I'm not too tired)

Sarah

**Reply by saraht45** [reply]
I am getting this error when compiling vault.c

> cd /usr/System/sys/tool
/usr/System/sys/tool
> ls
gcollect.c   patch.c      rebuild.c    vault.c
> compile vault.c

Failed to compile "/usr/System/sys/tool/vault.c"
     /kernel/obj/telnet#75827       'jominey'
  46     /kernel/obj/telnet             receive_message
 203     /kernel/lib/connection         receive_message
     /usr/System/obj/devuser#75924  'jominey'
 164     /usr/System/obj/devuser        receive_message
     /usr/System/obj/wiztool#124068
  38     /usr/System/obj/wiztool        input
 769     /kernel/lib/auto               call_limited
 127     /usr/System/obj/wiztool        process
 983     /kernel/lib/wiztool            cmd_compile
1021     /usr/System/lib/wiztool        compile_object
 256     /kernel/lib/wiztool            compile_object
 356     /kernel/lib/auto               compile_object
/usr/System/sys/tool/vault.c, 158: undeclared variable TLSD

**Reply by noahgibbs** [reply]
Hm... You could potentially install and load TLSD, but that may not be the best use of your time. If we're going to want some kind of drop-in XML export for older games then we'll want it to be pretty straightforward to use. TLSD was written recently by Shentino (now that I check) so none of the old games are going to have it. So it may be easier to modify the Vault Tool a bit to make a drop-in export tool.

Shentino is also right that there's probably some kind of existing XML export stuff related to Tree of WOE -- I've seen some things marked as object export but I don't yet know how to use them for the kind of export we're trying to do.

**Comment by shentino** [comment]
That smells like my code.

Where is the file hosted?

**Reply by noahgibbs** [reply]
At Chris's request, she's copying vaultd into an old statedump of Lovecraft Country in hopes of exporting the LPC objects as XML.

**Comment by shentino** [comment]
Also this code wasn't originally part of skotoslib proper IIRC and it may contain stuff that's set to automatically trigger on compilation.

I strongly advise not randomly compiling anything without at least a specific need or checking the code in question.

Anything I wrote there is likely to be dangerous if used improperly.

**Reply by ChristopherA** [reply]
I asked @saraht45 if she could backup to vault her copy of the Lovecraft Country objects.

**Reply by ChristopherA** [reply]
@shentino take a look at https://chattheatre.github.io/SkotOS-Doc/Exploring_SkotOS.html#dumping-lpc-objects-on-the-gables

**Reply by shentino** [reply]
@ChristopherA This seems outside the original intended use case for the vault dumper I wrote but from what I can tell the "can't dump, no state root" messages are intentional on my part because not every object findable through IDD (aka the woe tree) is dumpable due to the LPC source for the object in question not specifying a root SID object, since such object was apparently not intended to be configurable via woe or xml.  Please see reply on October 21, 9:36 pm pacific in this discussion for a fuller explanation.

**Comment by shentino** [comment]
Yeah, in order for an object to actually be configured in WOE it needs to have a registered SID object, whose path is returned by the LPC function query_state_root.  IIRC, not all objects registered in IDD with an object name are intended to be configurable via XML interfaces such as WOE.

More elaboratedly, SID appears to be the glue between the XML and the LPC, by among other things defining what LPC functions to call to import or export object state, and if an LPC source file doesn't specify a corresponding SID to sit at the top of the object state tree then it can't be manipulated in XML.

I have a hunch that the vault tool I wrote was an internal hack job for gables purely intended to dump the xml for all the objects, and IIRC it may even have preempted other tools that already existed, perhaps ones that generated xma files.

Honestly I'm the only one that's used the vault tool in production, back during my cold boot work for skotoslib, and I'm not really sure it's being used as intended here, and I regret that i never got around to writing documentation meant to guide other users.

This is also code I haven't touched in a long time, but I think the gist was to dump out xml for the objects for the purpose of cold booting.  I do not think it was meant as a backup strategy like the SID cemetery apparently was.

**Comment by shentino** [comment]
IIRC, the "can't dump" message is deliberate and meant to be a diagnostic indicating that a given object is intentionally being skipped for dumping because an SID root was never specified for the object in question.

**Comment by shentino** [comment]
I want to apologize generally if this code caused any problems but it was taken outside of its originally intended use and quite frankly for me even foreseeable use.

Please, whoever put this in should have either checked with me or read the source code or something before just slapping it in place.  This particular use case is outside the scope of the warranty, so to speak.

One other caveat that comes to mind that I should mention.  It appeared at the time that there were two different ways to dump an object, one with and without some kind of extra persistence, something to do with "transient" or something to do with data that was volatile and changed at runtime, perhaps things like inventory and so on that aren't meant to be part of the "blueprint" and IIRC the vault tool here only dumps that "baseline" version of the object.

I'd also like to point out that IIRC it probably doesn't dump anonymous spawns, such as cloned copies of a sword or a biscuit and so on.

All in all it's definitely NOT intended as some sort of game based backup tool, indeed.  At most it's intended to dump a library of named blueprints/masters/etc, as a sort of library or catalog you can spawn from.  If this was an attempt to use it to back up actual game/thing/world state data that is NOT what it was designed for at all, just in case.

**Reply by noahgibbs** [reply]
Hey, Shentino. I've actually read the code through, and I was the one who suggested this.

We're not trying to get a full backup of game state including where people were standing, anonymous spawns, etc. Chris was hoping we could get *only* the durable items (rooms, permanent items, etc.) so that we could eventually stand up another Lovecraft Country instance, no doubt with pretty careful checking as we do. It's not meant to back up a world-in-progress game state, just the important things like descriptions and areas that could potentially be imported into another game later.

So: the transient thing is no problem. In fact, it's intended behaviour from our point of view. Blueprints are exactly and specifically what we want here.

I don't intend any criticism about the error messages. As you say, they're intended behaviour and the tool isn't supposed to dump anything without a SID root. I wasn't suggesting the tool didn't work because of the error messages - just that she would see messages like that even if the tool *did* work as intended.

You're right, this wasn't the initial intended use. I should modify the vault tool a bit to reduce its dependencies, just to make it easier to use for this later. But I've read through the code and I think your tool as written is a pretty good match. We would really like to have drop-in code to back up LPC objects. And if we don't literally use your tool line-for-line as written for that, we may still use a modification of it. I *do* think it's a good match for this problem.

With that said, your earlier comment about using the WOE port for this may be right. That is, there may already be something built-in we can use.

**Reply by shentino** [reply]
I didn't say anything about the WOE port IIRC, just that the WOE interface works with object state exported via the SIDs, and the database of object names is stored in IDD, said database being where the folder tree on the left pane in WOE comes from.

My main concern is clarity, is all.  As for the error message no criticism was perceived, again I just wanted to be clear about what they meant and why I output them.

I'm glad it suits your purposes for the most part, it's what I used to create the object dump that was originally part of the open source release of SkotOS, namely the CC licensed object XML files.
